### PR TITLE
Implement mypy plugin for `luigi.Task`

### PR DESCRIPTION
**Pull request by hiro-o918**
* Add Mypy Plugin to type-check arguments of `luigi.Task`

## Description

* Implement a plugin for mypy inspired by [dataclass](https://github.com/python/mypy/blob/5dfc7d941253553ab77836e9845cb8fdfb9d23a9/mypy/plugins/dataclasses.py)
### Note
* bump up sphinx version and use Python 3.10 for docs environment 
  * because build failed on 3.9
  * [9b2e890](https://github.com/spotify/luigi/pull/3315/commits/9b2e89041047dc31e1e5042b80ebbbc52767395d)
* After writing tests, Python 3.12 failed on mypy errros.
  * [a0d8d2b](https://github.com/spotify/luigi/pull/3315/commits/a0d8d2bd78f9054a09245dcef846cb83f1024b8d)

## Motivation and Context

This PR aims to solve the following issues.

* the constructor of `luigi.Task` cannot be type-checked when initializing it.
* The following type annotation is invalid since `luigi.Parameter()` is not `str`
```
class Task:
  foo: str = luigi.Parameter()
```

## Have you tested this? If so, how?
 

* This PR includes the Mypy tests and docs

### CODE REVIEW

File: .github/workflows/pythonbuild.yml
Comment (hiro-o918) [inline .github/workflows/pythonbuild.yml:210]: use python 3.10 because docs build failed on 3.9
python for flake 8 also updated as a represented environment

File: luigi/mypy.py
Comment (dlstadther) [inline luigi/mypy.py:279]: Luigi still advertises to be 3.6 compatible. Though, support for 3.6 should be removed (readme reference, tox reference, github actions test execution, etc.).

Comment (dlstadther) [reply luigi/mypy.py:279]: I saw elsewhere in this PR that the mypy stuff requires 3.8 or later. What is the behavior when someone tries to use it with 3.6 or 3.7?

**Review by dlstadther — DISMISSED** [review]
LGTM. Just a mention/question about python versions

**Comment by dlstadther** [comment]
For some reason, the 3.9 version of docs and flake8 are being expected. Although I've approved the PR and am a maintainer of the project, I'm not being allowed to click merge due to 2 statuses missing.

File: luigi/mypy.py
Comment (hiro-o918) [reply luigi/mypy.py:279]: > I saw elsewhere in this PR that the mypy stuff requires 3.8 or later
Sorry about your confusing.
I made a mistake to set a type annotations like `list[Any]` which supported after 3.8
I fixed this issue, the tests worked on 3.6 and 3.7.

I'll check how it works on Python 3.6 via tests.
If the test become flaky, one of workaround is ommit supporting mypy plugion on 3.6

Comment (hiro-o918) [reply luigi/mypy.py:279]: After inspecting issues about python 3.6 and 3.7, some built-in features are lacking (e.g. some type annotation features like `Final` and `Literal`).
 
Can I want to omit these versions for this plugin because this feature is not core required to use luigi?

Comment (hiro-o918) [reply luigi/mypy.py:286]: [4744560](https://github.com/spotify/luigi/pull/3315/commits/474456006375343ea3de5a45a2873789ecd0f646)

Comment (dlstadther) [reply luigi/mypy.py:286]: Yeah, i'm fine with the included solution (requiring a minimum python version to import and use the mypy plugin).

**Review by dlstadther — APPROVED** [review]

**Comment by dlstadther** [comment]
To merge, we may require a Spotify employee who has the appropriate permissions. I think the CI is trying to wait on some of the tests expected by the `master` branch even though those same tests were updated to 3.10 and are running successfully.

@RRap0so , would you be able to look and possibly merge anyway? (I cannot merge, the button is grayed-out saying `Required statuses must pass before merging`).

**Comment by RRap0so** [comment]
@dlstadther all done 👍

**Pull request merged**
